src/modules/flow/update/update.c

```diff
@@ -0,0 +1,495 @@
+    }
+
+    return 0;
+}
+
+static struct sol_blob *
+serialize_info(const struct sol_update_info *response)
+{
+    struct sol_buffer buffer = SOL_BUFFER_INIT_EMPTY;
+    struct sol_blob *blob;
+    size_t size;
+    void *v;
+    int r;
+
+    r = sol_buffer_append_slice(&buffer, (const struct sol_str_slice)SOL_STR_SLICE_LITERAL("{\"version\":"));
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+    r = sol_json_serialize_string(&buffer, response->version);
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    r = sol_buffer_append_slice(&buffer, (const struct sol_str_slice)SOL_STR_SLICE_LITERAL(",\"url\":"));
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+    r = sol_json_serialize_string(&buffer, response->url);
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    r = sol_buffer_append_slice(&buffer, (const struct sol_str_slice)SOL_STR_SLICE_LITERAL(",\"hash\":"));
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+    r = sol_json_serialize_string(&buffer, response->hash);
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    r = sol_buffer_append_slice(&buffer, (const struct sol_str_slice)SOL_STR_SLICE_LITERAL(",\"hash-algorithm\":"));
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+    r = sol_json_serialize_string(&buffer, response->hash_algorithm);
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    r = sol_buffer_append_char(&buffer, '}');
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    v = sol_buffer_steal(&buffer, &size);
+    blob = sol_blob_new(SOL_BLOB_TYPE_DEFAULT, NULL, v, size);
+    SOL_NULL_CHECK_GOTO(blob, err);
+
+    return blob;
+err:
+    sol_buffer_fini(&buffer);
+
+    return NULL;
+}
+
+static void
+check_cb(void *data, int status, const struct sol_update_info *response)
+{
+    struct base_data *mdata = data;
+    struct sol_blob *blob;
+
+    if (status < 0) {
+        sol_flow_send_error_packet(mdata->node, -status,
+            "Error while checking for updates");
+        goto end;
+    }
+
+    /* TODO maybe a custom packet instead of JSON? Maybe sol_update_info have
+     * the raw json? */
+    blob = serialize_info(response);
+    if (blob) {
+        sol_flow_send_json_object_packet(mdata->node,
+            SOL_FLOW_NODE_TYPE_UPDATE_CHECK__OUT__UPDATE_INFO, blob);
+        sol_blob_unref(blob);
+    } else {
+        sol_flow_send_error_packet(mdata->node, EINVAL,
+            "Could not create JSON for UPDATE_INFO");
+    }
+
+    sol_flow_send_string_packet(mdata->node,
+        SOL_FLOW_NODE_TYPE_UPDATE_CHECK__OUT__VERSION, response->version);
+
+end:
+    mdata->handle = NULL;
+}
+
+static int
+check_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct check_data *mdata = data;
+
+    if (!mdata->url) {
+        SOL_WRN("No URL to check for updates");
+        return -EINVAL;
+    }
+
+    if (mdata->base.handle) {
+        sol_flow_send_error_packet(node, EINVAL,
+            "Check already in progress. You may want to CANCEL before issuing a new CHECK");
```

Comment:
I don't see why we should fail. Just ignore, if there is a check, then he will get a packet at some point.